listenbrainz/webserver/static/js/src/BrainzPlayer.tsx

```diff
@@ -0,0 +1,403 @@
+import * as React from "react";
+import { isEqual as _isEqual, get as _get } from "lodash";
+import * as _ from "lodash";
+import PlaybackControls from "./PlaybackControls";
+import APIService from "./APIService";
+import SpotifyPlayer from "./SpotifyPlayer";
+import YoutubePlayer from "./YoutubePlayer";
+
+export type DataSourceType = {
+  playListen: (listen: Listen) => void;
+  togglePlay: () => void;
+  seekToPositionMs: (msTimecode: number) => void;
+};
+
+export type DataSourceTypes = SpotifyPlayer | YoutubePlayer;
+
+export type DataSourceProps = {
+  show: boolean;
+  playerPaused: boolean;
+  onPlayerPausedChange: (paused: boolean) => void;
+  onProgressChange: (progressMs: number) => void;
+  onDurationChange: (durationMs: number) => void;
+  onTrackInfoChange: (title: string, artist?: string) => void;
+  onTrackEnd: () => void;
+  onTrackNotFound: () => void;
+  handleError: (error: string | Error, title?: string) => void;
+  handleWarning: (message: string | JSX.Element, title?: string) => void;
+  handleSuccess: (message: string | JSX.Element, title?: string) => void;
+  onInvalidateDataSource: (
+    dataSource?: DataSourceTypes,
+    message?: string | JSX.Element
+  ) => void;
+};
+
+type BrainzPlayerProps = {
```

Comment:
I wonder if we should consider adding to the props the sources we want brainzplayer to use, in priority order. Maybe not right now though, once we find a usecase.

Reply:
That's definitely on my mind.
The way I see it, at some point we'll need an overhaul of what is currently the connect-spotify page to allow for other music services we want to connect to.

Along with that, we'll want somewhere in the user profile a list of the services the user is connected to and a way to arrange them by order of preference.
Of course after that the BrainzPlayer will need to be modified to take that order into account.